File: libmuscle/python/libmuscle/operator.py
from typing import Dict

from ymmsl import Operator
import muscle_manager_protocol.muscle_manager_protocol_pb2 as mmp


def operator_from_grpc(
        operator: mmp.Operator
        ) -> Operator:
    """Creates an operator from a gRPC-generated Operator.

    Args:
        operator: An operator, received from gRPC.

    Returns:
        The same operator, as an Operator.
    """
    operator_map = {
            mmp.OPERATOR_NONE: Operator.NONE,
            mmp.OPERATOR_F_INIT: Operator.F_INIT,
            mmp.OPERATOR_O_I: Operator.O_I,
            mmp.OPERATOR_S: Operator.S,
            mmp.OPERATOR_B: Operator.B,
            mmp.OPERATOR_O_F: Operator.O_F
            }   # type: Dict[int, Operator]
    return operator_map[operator]


def operator_to_grpc(operator: Operator) -> mmp.Operator:
    """Converts the operator to the gRPC generated type.

    Returns:
        The operator, as the gRPC type.
    """
    operator_map = {
            Operator.NONE: mmp.OPERATOR_NONE,
            Operator.F_INIT: mmp.OPERATOR_F_INIT,
            Operator.O_I: mmp.OPERATOR_O_I,
            Operator.S: mmp.OPERATOR_S,
            Operator.B: mmp.OPERATOR_B,
            Operator.O_F: mmp.OPERATOR_O_F
            }   # type: Dict[Operator, mmp.Operator]
    return operator_map[operator]

File: libmuscle/cpp/src/libmuscle/operator.cpp
#include "operator.hpp"


namespace mmp = muscle_manager_protocol;


namespace libmuscle { namespace impl {

ymmsl::Operator operator_from_grpc(mmp::Operator op) {
    ymmsl::Operator result;

    switch (op) {
        case mmp::OPERATOR_NONE:
            result = ymmsl::Operator::NONE;
            break;
        case mmp::OPERATOR_F_INIT:
            result = ymmsl::Operator::F_INIT;
            break;
        case mmp::OPERATOR_O_I:
            result = ymmsl::Operator::O_I;
            break;
        case mmp::OPERATOR_S:
            result = ymmsl::Operator::S;
            break;
        case mmp::OPERATOR_B:
            result = ymmsl::Operator::B;
            break;
        case mmp::OPERATOR_O_F:
            result = ymmsl::Operator::O_F;
            break;
        case mmp::OPERATOR_MAP:
            result = ymmsl::Operator::NONE;
            break;
        default:
            // https://www.xkcd.com/2200
            throw std::runtime_error(
                    "ERROR: We've reached an unreachable state."
                    " Anything is possible."
                    " The limits were in our heads all along."
                    " Follow your dreams.");
    }
    return result;
}

muscle_manager_protocol::Operator operator_to_grpc(ymmsl::Operator op) {
    mmp::Operator result;

    switch (op) {
        case ymmsl::Operator::NONE:
            result = mmp::OPERATOR_NONE;
            break;
        case ymmsl::Operator::F_INIT:
            result = mmp::OPERATOR_F_INIT;
            break;
        case ymmsl::Operator::O_I:
            result = mmp::OPERATOR_O_I;
            break;
        case ymmsl::Operator::S:
            result = mmp::OPERATOR_S;
            break;
        case ymmsl::Operator::B:
            result = mmp::OPERATOR_B;
            break;
        case ymmsl::Operator::O_F:
            result = mmp::OPERATOR_O_F;
            break;
    }
    return result;
}

} }

